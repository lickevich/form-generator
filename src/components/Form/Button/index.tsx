import React from "react";

import { StyledInput } from "./style";
import { IButton } from "../../../type";

const Button: React.FC<IButton> = ({ type, value, ...rest }) => (
  <StyledInput type={type} value={value} {...rest} />
);

export default Button;
