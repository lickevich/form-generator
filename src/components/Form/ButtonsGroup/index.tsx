import React from "react";

import { StyledButtonsGroup } from "./style";

const ButtonsGroup: React.FC<React.ReactNode> = ({ children }) => (
  <StyledButtonsGroup>{children}</StyledButtonsGroup>
);

export default ButtonsGroup;
