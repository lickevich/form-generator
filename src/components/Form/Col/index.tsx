import React from "react";

import { StyledCol } from "./style";

const Col: React.FC<React.ReactNode> = ({ children }) => (
  <StyledCol>{children}</StyledCol>
);

export default Col;
