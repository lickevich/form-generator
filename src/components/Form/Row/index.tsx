import React from "react";

import { StyledRow } from "./style";

const Row: React.FC<React.ReactNode> = ({ children }) => (
  <StyledRow>{children}</StyledRow>
);

export default Row;
