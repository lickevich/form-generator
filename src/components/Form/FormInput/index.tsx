import React from "react";

import { IInput } from "../../../type";
import Row from "../Row";
import Col from "../Col";

const FormInput: React.FC<IInput> = ({ index, label, ...rest }) => {
  const id = `${label}${index}`;

  return (
    <Row>
      <Col>
        <label htmlFor={id} title={label}>
          {label}
        </label>
      </Col>
      <Col>
        <input id={id} {...rest} />
      </Col>
    </Row>
  );
};

export default FormInput;
