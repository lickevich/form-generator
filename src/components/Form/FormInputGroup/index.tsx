import React from "react";

import { IInput } from "../../../type";
import Row from "../Row";
import Col from "../Col";

const FormInputGroup: React.FC<IInput> = ({
  type = "",
  groupLabel = "",
  items = [],
}) => {
  return (
    <Row>
      <Col>
        <label title={groupLabel}>{groupLabel}</label>
      </Col>
      <Col>
        {items.map(({ label = "", ...rest }, index) => {
          const id = `${label}${index}`;

          return (
            <div key={id}>
              <input id={id} type={type} {...rest} />
              <label htmlFor={id} title={label}>
                {label}
              </label>
            </div>
          );
        })}
      </Col>
    </Row>
  );
};

export default FormInputGroup;
