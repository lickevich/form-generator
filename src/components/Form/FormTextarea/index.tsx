import React from "react";

import { IInput } from "../../../type";
import Row from "../Row";
import Col from "../Col";

const FormTextarea: React.FC<IInput> = ({
  index = "",
  label = "",
  ...rest
}) => {
  const id = `${label}${index}`;

  return (
    <Row>
      <Col>
        <label htmlFor={id} title={label}>
          {label}
        </label>
      </Col>
      <Col>
        <textarea id={id} {...rest} />
      </Col>
    </Row>
  );
};

export default FormTextarea;
