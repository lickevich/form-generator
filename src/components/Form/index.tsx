import React, { Fragment } from "react";

import { ITemplate } from "../../type";
import { Title, StyledForm } from "./style";
import {
  inputTypes,
  isCheckboxType,
  isRadioType,
  isTextareaType,
} from "../../models";
import { isIncludedValue } from "../../utils/helpers";
import FormInput from "../Form/FormInput";
import FormTextarea from "../Form/FormTextarea";
import FormInputGroup from "../Form/FormInputGroup";
import ButtonsGroup from "./ButtonsGroup";
import Button from "./Button";
import Row from "./Row";

const Form: React.FC<ITemplate> = ({ title, items, buttons, ...rest }) => {
  return (
    <Fragment>
      <Row>
        <Title>{title}</Title>
      </Row>
      <StyledForm {...rest}>
        {items.map(
          ({ type = "", label = "", groupLabel = "", ...rest }, index) => {
            const isInputTypes: boolean = isIncludedValue(inputTypes, type);
            const id: string = `${label}${index}`;

            if (isInputTypes) {
              return <FormInput key={id} type={type} label={label} {...rest} />;
            }
            if (isTextareaType(type)) {
              return (
                <FormTextarea key={id} type={type} label={label} {...rest} />
              );
            }
            if (isCheckboxType(type) || isRadioType(type)) {
              return (
                <FormInputGroup
                  key={id}
                  type={type}
                  groupLabel={groupLabel}
                  {...rest}
                />
              );
            }

            return null;
          }
        )}
      </StyledForm>
      <ButtonsGroup>
        {buttons.map((button, index) => {
          const id: string = `${button.value}${index}`;
          return <Button key={id} {...button} />;
        })}
      </ButtonsGroup>
    </Fragment>
  );
};

export default Form;
