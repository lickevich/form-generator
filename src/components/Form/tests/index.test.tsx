import React from "react";
import { render } from "@testing-library/react";

import Form from "../index";

describe("Form", () => {
  const template = {
    title: "Form Title",
    items: [
      { type: "text", label: "textfield" },
      { type: "number", label: "numberfield" },
      { type: "textarea", label: "textarea" },
      { type: "date", label: "datefield" },
      {
        type: "checkbox",
        groupLabel: "checkbox",
        items: [{ label: "checkbox1" }, { label: "checkbox2" }],
      },
      {
        type: "radio",
        groupLabel: "radio",
        items: [
          { label: "radio1", name: "radio", value: "1" },
          { label: "radio2", name: "radio", value: "2" },
        ],
      },
    ],
    buttons: [
      { type: "submit", value: "Apply" },
      { type: "button", value: "Ok" },
      { type: "reset", value: "Reset" },
    ],
  };

  it("render the component", () => {
    const component = <Form {...template} />;
    const { asFragment } = render(component);

    // @ts-ignore
    expect(asFragment(component)).toMatchSnapshot();
  });
});
