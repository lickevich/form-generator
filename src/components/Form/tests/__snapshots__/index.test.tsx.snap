// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form render the component 1`] = `
<DocumentFragment>
  <div
    class="sc-dkPtyc jEDgkp"
  >
    <h1
      class="sc-bdvvaa bDsQHB"
    >
      Form Title
    </h1>
  </div>
  <form
    class="sc-gsDJrp hjUSrA"
  >
    <div
      class="sc-dkPtyc jEDgkp"
    >
      <div
        class="sc-hKwCoD eERlPP"
      >
        <label
          for="textfieldundefined"
          title="textfield"
        >
          textfield
        </label>
      </div>
      <div
        class="sc-hKwCoD eERlPP"
      >
        <input
          id="textfieldundefined"
          type="text"
        />
      </div>
    </div>
    <div
      class="sc-dkPtyc jEDgkp"
    >
      <div
        class="sc-hKwCoD eERlPP"
      >
        <label
          for="numberfieldundefined"
          title="numberfield"
        >
          numberfield
        </label>
      </div>
      <div
        class="sc-hKwCoD eERlPP"
      >
        <input
          id="numberfieldundefined"
          type="number"
        />
      </div>
    </div>
    <div
      class="sc-dkPtyc jEDgkp"
    >
      <div
        class="sc-hKwCoD eERlPP"
      >
        <label
          for="textarea"
          title="textarea"
        >
          textarea
        </label>
      </div>
      <div
        class="sc-hKwCoD eERlPP"
      >
        <textarea
          id="textarea"
          type="textarea"
        />
      </div>
    </div>
    <div
      class="sc-dkPtyc jEDgkp"
    >
      <div
        class="sc-hKwCoD eERlPP"
      >
        <label
          for="datefieldundefined"
          title="datefield"
        >
          datefield
        </label>
      </div>
      <div
        class="sc-hKwCoD eERlPP"
      >
        <input
          id="datefieldundefined"
          type="date"
        />
      </div>
    </div>
    <div
      class="sc-dkPtyc jEDgkp"
    >
      <div
        class="sc-hKwCoD eERlPP"
      >
        <label
          title="checkbox"
        >
          checkbox
        </label>
      </div>
      <div
        class="sc-hKwCoD eERlPP"
      >
        <div>
          <input
            id="checkbox10"
            type="checkbox"
          />
          <label
            for="checkbox10"
            title="checkbox1"
          >
            checkbox1
          </label>
        </div>
        <div>
          <input
            id="checkbox21"
            type="checkbox"
          />
          <label
            for="checkbox21"
            title="checkbox2"
          >
            checkbox2
          </label>
        </div>
      </div>
    </div>
    <div
      class="sc-dkPtyc jEDgkp"
    >
      <div
        class="sc-hKwCoD eERlPP"
      >
        <label
          title="radio"
        >
          radio
        </label>
      </div>
      <div
        class="sc-hKwCoD eERlPP"
      >
        <div>
          <input
            id="radio10"
            name="radio"
            type="radio"
            value="1"
          />
          <label
            for="radio10"
            title="radio1"
          >
            radio1
          </label>
        </div>
        <div>
          <input
            id="radio21"
            name="radio"
            type="radio"
            value="2"
          />
          <label
            for="radio21"
            title="radio2"
          >
            radio2
          </label>
        </div>
      </div>
    </div>
  </form>
  <div
    class="sc-eCImvq UFQQy"
  >
    <input
      class="sc-jRQAMF EJfAC"
      type="submit"
      value="Apply"
    />
    <input
      class="sc-jRQAMF EJfAC"
      type="button"
      value="Ok"
    />
    <input
      class="sc-jRQAMF EJfAC"
      type="reset"
      value="Reset"
    />
  </div>
</DocumentFragment>
`;
